File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        r=len(grid)
        c=len(grid[0])
        def valid(row,col):
            return 0<=row<r and 0<=col<c and grid[row][col]=='1'
        directions=[(0,1),(1,0),(-1,0),(0,-1)]
        seen=set()
        cnt=0
        def dfs(row,col):
            for dx,dy in directions:
                newRow=row+dx
                newCol=col+dy
                if valid(newRow,newCol) and (newRow,newCol) not in seen:
                    seen.add((newRow,newCol))
                    dfs(newRow,newCol)

        for i in range(r):
            for j in range(c):
                if (i,j) not in seen and valid(i,j):
                    seen.add((i,j))
                    cnt+=1
                    dfs(i,j)

        return cnt

File: Data Structures & Algorithms/islands-and-treasure/submission-5.java
class Solution {
    boolean[][] seen;
    public boolean valid(int[][] grid,int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return (0<=row)&&(row<r)&&(0<=col)&&(col<c)&&(grid[row][col]!=-1);
    }
    public void islandsAndTreasure(int[][] grid) {
        int r=grid.length;
        int c=grid[0].length;
        Deque<int[]> queue=new ArrayDeque<>();
        seen=new boolean[r][c];
        int[][] directions=new int[][]{{0,1},{1,0},{-1,0},{0,-1}};
        for (int i=0;i<r;i++){
            for (int j=0;j<c;j++){
                if (grid[i][j]==0){
                    seen[i][j]=true;
                    queue.offer(new int[] {i,j});
                }
            }
        }
        int traverse=1;
        while (!queue.isEmpty()){
            int len=queue.size();
            for(int i=0;i<len;i++){
                int[] curr=queue.poll();
                int row=curr[0];
                int col=curr[1];
                for (int[] dir:directions){
                    int newRow=row+dir[0];
                    int newCol=col+dir[1];
                    if (valid(grid,newRow,newCol)&&(seen[newRow][newCol]==false)){
                        seen[newRow][newCol]=true;
                        grid[newRow][newCol]=traverse;
                        queue.offer(new int[]{newRow,newCol});
                    }
                }
            }
            traverse+=1;
        }
        
    }
}
